icefoganalytics/internal-data-portal
# Build Dataset Access Subscriptions Email Users Button

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/41

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- https://github.com/icefoganalytics/internal-data-portal/issues/38
- https://github.com/icefoganalytics/internal-data-portal/issues/90

# Context

Add Email users action, and ability to send emails.

Example Mailers 
- https://github.com/icefoganalytics/signing-authority/tree/main/api
- https://github.com/icefoganalytics/signing-authority/blob/main/api/src/services/email-service.ts
- https://github.com/icefoganalytics/signing-authority/tree/main/api/src/templates/email
- https://github.com/icefoganalytics/signing-authority/blob/main/api/src/config.ts

## Dataset Access Page - /datasets/:slug/access

### Front-end routes

Only the page route, though one could argue that the nearby tabs should be included. Note of the components on this page redirect away from it, they just open modals

### Back-end routes

* /api/email-templates/:slug/new?data_set_id=.. - virtual model, for now, generates the prefilled content of notify users email
* /api/emails - POST, enqueue a new email job

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue` - see previous definition
* `DatasetAccessPage.vue` - only available to higher rank users
    * `SubscriptionsCard.vue`
        * `EmailUsersButton.vue` - opens email subscribers dialog, prefils data from /emails/new?data_set_id=...? Or maybe /datasets/email-subscribers/new ? or maybe /emails/dataset-template/new?data_set_id=...? 
        * `EmailSubscribersDialog.vue`
            * Email - prefilled, but editable, list of emails from all users with access requests or subscriptions to this dataset.
            * Title - prefilled, but editable, email title
            * Content - text input for email body, in future this should support rich text or markdown, but that isn’t specified in the wireframe
            * Cancel - cancel button
            * Send - send email button, POST to /emails

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/3df034b2-899f-43b1-8fb2-5a17b057dd80)

# Implementation

Add /datasets/:slugOrId/email-subsribers controller and endpoint
Add BaseMailer that handle render logic and simplifies logic for individual mailers.
Store mailer templates in /templates folder.
Use simple `<pre>` tag to render content as formatted text. Keeps PR simple until https://github.com/icefoganalytics/internal-data-portal/issues/90 can be implemented.
Use MailDev docker compose setup for easy development testing of email.

# Screenshots

Prefilled email user dialog
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/199bde84-8834-4012-80ab-a847ec490f8a)

Email as seen by MailDev
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/4d238e5b-f5d1-4025-9d4b-3b30e2231cba)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Open http://localhost:1080 to see the new MailDev mail server setup.
4. In a second tab, log in to the app at http://localhost:8080.
5. Create a dataset with some limited access grants.
6. Log in as some other users and subscribe to the dataset.
7. Log back in as the data owner and go to the manage access tab of the dataset.
8. You will see the available subscriptions.
9. Click on the "Email Users" button in the "Subscriptions" panel to open the "Email Users" dialog.
10. Note that the list of emails is pre-filled from the list of subscribers.
11. Note that you can add any extra emails you need to the "Email" section.
12. Note that the title is prefilled from the dataset title.
13. Add some content to the content section and update the title as needed.
    e.g.
    ```
    Dear Subscribers,
    
    The Internal Data Portal now has the ability to email subscribers.
    
    Regards,
        Some Person
    ```

14. Note the formatting of the text, it will carry through when the email is sent.
15. Click "send" to send the email.
16. Note that the "MailDev" tab will show up as having a new email; switch to that tab to read the email.
17. Click on the new unread email to display the email with formatting preserved.
18. Check that the "text" version of the email also displays correctly; it is the raw content of the "content" section.
